pkg/mimirtool/commands/remote_read.go

```diff
@@ -247,36 +257,353 @@ func (c *RemoteReadCommand) prepare() (query func(context.Context) (storage.Seri
+}
+
+func (c *combinedSeriesSet) At() storage.Series {
+	if c.index < 0 || c.index >= len(c.series) {
+		return nil
+	}
+	return c.series[c.index]
+}
+
+func (c *combinedSeriesSet) Err() error {
+	return c.err
+}
+
+func (c *combinedSeriesSet) Warnings() annotations.Annotations {
+	return nil
+}
+
+// handleSampledResponse handles the traditional sampled response format
+func (c *RemoteReadCommand) handleSampledResponse(httpResp *http.Response, queries []*prompb.Query) (storage.SeriesSet, error) {
+	// Read and decompress response
+	compressedResp, err := io.ReadAll(httpResp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("error reading response: %w", err)
+	}
+
+	uncompressed, err := snappy.Decode(nil, compressedResp)
+	if err != nil {
+		return nil, fmt.Errorf("error decompressing response: %w", err)
+	}
+
+	// Unmarshal response
+	var resp prompb.ReadResponse
+	err = proto.Unmarshal(uncompressed, &resp)
+	if err != nil {
+		return nil, fmt.Errorf("unable to unmarshal response body: %w", err)
+	}
+
+	if len(resp.Results) != len(queries) {
+		return nil, fmt.Errorf("responses: want %d, got %d", len(queries), len(resp.Results))
+	}
+
+	// Combine all results from all queries
+	var allSeries []storage.Series
+	for i, result := range resp.Results {
+		log.Infof("Processing result %d/%d with %d series", i+1, len(resp.Results), len(result.Timeseries))
+		seriesSet := remote.FromQueryResult(false, result)
+		for seriesSet.Next() {
+			allSeries = append(allSeries, seriesSet.At())
+		}
+		if err := seriesSet.Err(); err != nil {
+			return nil, fmt.Errorf("error reading series from query %d: %w", i, err)
+		}
+	}
+
+	log.Infof("Combined %d series from %d queries", len(allSeries), len(queries))
+	return &combinedSeriesSet{series: allSeries, index: -1}, nil
+}
+
+// handleChunkedResponse handles the streamed chunked response format
+func (c *RemoteReadCommand) handleChunkedResponse(httpResp *http.Response, queries []*prompb.Query) (storage.SeriesSet, error) {
+	// Use the chunked reader from the remote package
+	reader := remote.NewChunkedReader(httpResp.Body, c.readSizeLimit, nil)
+
+	// Collect all series from all queries
+	var allSeries []storage.Series
+	processedQueries := make(map[int64]int)
+	totalBytes := 0
+
+	for {
+		var chunkedResp prompb.ChunkedReadResponse
+		err := reader.NextProto(&chunkedResp)
+		if errors.Is(err, io.EOF) {
+			break
+		}
+		totalBytes += chunkedResp.Size()
 		if err != nil {
-			return nil, err
+			return nil, fmt.Errorf("error reading chunked response: %w", err)
 		}
 
-		return resp, nil
+		queryIndex := chunkedResp.QueryIndex
+		if queryIndex < 0 || queryIndex >= int64(len(queries)) {
+			return nil, fmt.Errorf("invalid query index %d, expected 0-%d", queryIndex, len(queries)-1)
+		}
 
-	}, from, to, nil
+		processedQueries[queryIndex]++
+		if processedQueries[queryIndex] == 1 {
+			log.Infof("Processing chunked response for query %d", queryIndex)
+		}
+
+		// Each ChunkedSeries contains chunks for a single series
```

Comment:
Can't a single series be spread across multiple consecutive protobuf messages?

https://github.com/prometheus/prometheus/blob/4aee718013c52bef1ccb336bcc21886336e6252e/prompb/remote.proto#L80-L82 says:

```
// We strictly stream full series after series, optionally split by time. This means that a single frame can contain
// partition of the single series, but once a new series is started to be streamed it means that no more chunks will
// be sent for previous one. Series are returned sorted in the same way TSDB block are internally.
```